labs/KeyWrapping/KeyWrapping.md

```diff
@@ -0,0 +1,1015 @@
+    C      = zero
+```
+
+With `W` in hand there it isn't much more work to complete the `KW-AE`
+algorithm.
+
+**EXERCISE**: Study `Algorithm 3` and complete the specification for
+  `KW-AE` by filling in the snippet below with the correct definition
+  for `S`. Remember that `ICV1` has already been defined in this
+  module. Also, notice that the bounds from Table 1 (Section 5.3.1,
+  page 10) are included as type constraints.
+
+```
+KWAE :
+    {n}
+    (fin n, 2 <= n, n < 2^^54) =>
+    ([128] -> [128]) -> [n][64] -> [n+1][64]
+KWAE CIPHk P = C
+  where
+    S = zero : [n+1][64]
+    C = zero
+```
+
+At this point you can check your work against six test vectors given
+in a property defined later on in this document. Here is the the
+command and sample output for `KWAETests`.
+
+```shell
+Cryptol> :check KWAETests
+Using exhaustive testing.
+Passed 1 tests.
+Q.E.D.
+```
+
+## Building a Formal Specification for `KW-AD`
+
+It should be fairly straightforward at this point to implement the
+authenticated decryption function `KW-AD` which is the other algorithm
+in the `KW` family. There is one significant difference, though: since
+the algorithm *authenticates* the decryption we will need to add an
+authentication check and slightly alter the type for `KW-AD`.
+
+**EXERCISE**: Review `Algorithm 2` and `Figure 3` of the document and
+complete the definitions for the inverse routines to `W` and `WStep`
+which we shall call `W'` and `WStep'` respectively. The type
+declarations for these functions are provided.
+
+*Hint:* Notice that, except for the names, the type declarations are
+identical. The function definitions are also very similar. Pay
+special attention to the order of the index variable for the main
+loop, the sequence of operations, and how the sequence of `Rs`
+transforms:
+
+```
+WStep' :
+    {n}
+    (fin n, n >= 3) =>
+    ([128] -> [128]) -> [n][64] -> [64] -> [n][64]
+WStep' CIPHk' ([A] # Rs) t = [A'] # Rs'
+  where
+    [MSB, LSB] = zero : [2][64]
+    A'         = zero
+    Rs'        = zero
+
+W' :
+    {n}
+    (fin n, 3 <= n, n <= 2^^54) =>
+    ([128] -> [128]) -> [n][64] -> [n][64]
+W' CIPHk' C = S
+  where
+    type s = 0
+    ts     = zero : [s][64]
+    S      = zero
+```
+
+Once you have these completed you should be able to check your work by
+having Cryptol `:prove` the properties `WStep'Prop` and
+`W'Prop`. Your output should look something like the following:
+
+```shell
+Cryptol> :prove WStep'Prop 
+Q.E.D.
+(Total Elapsed Time: 0.079s, using Z3)
+Cryptol> :prove W'Prop
+Q.E.D.
+(Total Elapsed Time: 1.832s, using Z3)
+```
+
+These two properties state that for a fixed, dummy CIPHk and S of
+length 3 semiblocks, `WStep` and `WStep'` are inverses and `W` and
+`W'` are inverses. Here are the definitions of these properties:
+
+```
+property WStep'Prop ARs t =
+    WStep'`{3} (\a -> a-1) (WStep (\a -> a+1) ARs t) t == ARs
+
+property W'Prop S =
+    W'`{3} (\a -> a-1) (W (\a -> a+1) S) == S
+```
+
```

Comment:
They are just dummy functions that are inverses of eachother. We're really supposed to be using AES here, but the prover will take too long because it doesn't understand apriori that AESencrypt and AESdecrypt are inverses...